modules/database/src/main/java/com/intuit/wasabi/database/impl/DBITransaction.java

```diff
@@ -178,78 +197,82 @@ private Object perform(HandleBlock block) {
 
             throw ex;
         } finally {
-            if (!inTransaction) {
-                close();
+            // Close the handle (internally this call releases the JDBC connection to the pool)
+            if (handle!=null) {
+                handle.close();
             }
         }
     }
 
+    /**
+     * Creates a new handle to perform database operation.
+     * Handle represents a connection to the database system. It is a wrapper around a JDBC Connection object.
+     *
+     * @return handle to perform database operation.
+     *
+     */
     protected Handle getHandle() {
-        if (null == handle) {
-            handle = dbi.open();
-            // Force MySQL timezone to UTC
-            handle.execute("set time_zone = \"+0:00\"");
-        }
+        //This operation reuses JDBC Connection object from the pool and then creates a new handle object.
+        Handle handle = dbi.open();
+
+        // Force MySQL timezone to UTC
+        handle.execute("set time_zone = \"+0:00\"");
 
         return handle;
     }
 
-    protected void begin() throws WasabiServerException {
-        Handle h = getHandle();
-
+    /**
+     * Begin a new transaction for this instance of DBITransaction.
+     *
+     * Initialize the handle of this instance of DBITransaction
+     *
+     * @throws WasabiServerException
+     */
+    protected void begin(Handle handle) throws WasabiServerException {
```

Comment:
DatabaseException is specific type of WasabiServerException... So it is good idea to have  WasabiServerException in throws clause.